python/strings_udf/cpp/include/cudf/strings/udf/udf_string.cuh

```diff
@@ -0,0 +1,475 @@
+/*
+ * Copyright (c) 2020-2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "udf_string.hpp"
+
+#include <cudf/strings/detail/utf8.hpp>
+#include <cudf/strings/string_view.cuh>
+
+#include <algorithm>
+#include <limits>
+#include <string>
+
+namespace cudf {
+namespace strings {
+namespace udf {
+namespace detail {
+
+__device__ inline static cudf::size_type bytes_in_null_terminated_string(char const* str)
+{
+  if (!str) return 0;
+  cudf::size_type bytes = 0;
+  while (*str++)
+    ++bytes;
+  return bytes;
+}
+
+}  // namespace detail
+
+__device__ inline char* udf_string::allocate(cudf::size_type bytes)
+{
+  char* data  = static_cast<char*>(malloc(bytes + 1));
+  data[bytes] = 0;  // add null-terminator so we can printf strings in device code
+  return data;
+}
+
+__device__ inline void udf_string::deallocate(char* data)
```

Comment:
This can be called with `data != this->mdata`. I do not agree with making it a static method. It is a specialized method for this class and likely in the future require access to specific data members (i.e. custom allocator like for `std::string`).

Reply:
Discussed offline, I'm OK with leaving this as is.